actors/builtin/market/testing.go

```diff
@@ -0,0 +1,229 @@
+package market
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/ipfs/go-cid"
+	"github.com/pkg/errors"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin"
+	"github.com/filecoin-project/specs-actors/v2/actors/util/adt"
+)
+
+type StateSummary struct {
+	ProposalCount        uint64
+	PendingProposalCount uint64
+	DealStateCount       uint64
+	LockTableCount       uint64
+	DealOpEpochCount     uint64
+	DealOpCount          uint64
+}
+
+// Checks internal invariants of market state.
+func CheckStateInvariants(st *State, store adt.Store) (*StateSummary, *builtin.MessageAccumulator, error) {
+	acc := &builtin.MessageAccumulator{}
+
+	acc.Require(
+		st.TotalClientLockedCollateral.GreaterThanEqual(big.Zero()),
+		"negative total client locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalProviderLockedCollateral.GreaterThanEqual(big.Zero()),
+		"negative total provider locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalClientStorageFee.GreaterThanEqual(big.Zero()),
+		"negative total client storage fee: %v", st.TotalClientLockedCollateral)
+
+	//
+	// Proposals
+	//
+
+	allIDs := make(map[abi.DealID]struct{})
+	proposalCids := make(map[cid.Cid]struct{})
+	maxDealID := int64(-1)
+	proposalCount := uint64(0)
+
+	proposals, err := adt.AsArray(store, st.Proposals)
+	if err != nil {
+		return nil, acc, err
+	}
+	var proposal DealProposal
+	err = proposals.ForEach(&proposal, func(dealID int64) error {
+		allIDs[abi.DealID(dealID)] = struct{}{}
+
+		pcid, err := proposal.Cid()
+		if err != nil {
+			return err
+		}
+
+		// keep some state
+		proposalCids[pcid] = struct{}{}
+		if dealID > maxDealID {
+			maxDealID = dealID
+		}
+		proposalCount++
+
+		return nil
```

Comment:
I think there's something about proposal cleanup, where every proposal should either have an associated state, or be within `DealUpdatesInterval` of it's start date (about to be expired). Proposals more than `DealUpdatesInterval` after their EndDate should not exist, and neither should associated state. From `CronTick`.

The Proposal/State/OpEpoch inter-consistency might require loading all of those into Go maps for easier lookups.

Reply:
I've added a map to track proposals expected to have an op and ensure that everything in that map has one.